framework/filters/AccessRule.php

```diff
@@ -161,8 +191,13 @@ protected function matchRole($user)
                 if (!$user->getIsGuest()) {
                     return true;
                 }
-            } elseif ($user->can($role)) {
-                return true;
+            } else {
+                if (!isset($roleParams)) {
```

Comment:
I moved this into the foreach to evaluate it only if it is needed, e.g. it will not evaluate if '?' or '@' rule will return true first.